Optimizing pv_stpv and EQE_spectral_response

AnonymousOceans opened the pull request:
## Description
Added self consistent temperature calculation to compute_pv_stpv, as well as a short circuit computation for compute_pv_stpv, and a calculation for EQE spectral response, with data.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - Implement self-consistent temperature calculation to balance absorptivity and emissivity
  - Fixed docstrings for compute_pv_stpv, compute_pv_stpv_gradient, and EQE_spectral_response functions
  - Implemented short circuit current calculation for compute_pv_stpv
  - Added argument for EQE_spectral_response under em.py

## Questions
- Do we need unit testing for the temperature calculation?

## Status
- Code Review

AnonymousOceans committed:
Added EQE spectral response, compute pv/stpv, and compute pv/stpv short circuit current functions.

FoleyLab commented on wptherml/em.py:
Follow the if statement (lines 318/319) with an else statement that will set a default value for self.psc_thickness_option if the user does not include this flag in their input arguments.

FoleyLab replied:
The compute_pv_stpv() method on line 968 should resemble the structure of the compute_stpv() method on around 944, that is it should just call the various methods that compute the specific figures of merit for pv_stpv (pv_stpv_jsc, pv_stpv_splitting_power for now).  It can also call the self-consistent temperature, but right now I commented out the loop within the self-consistent temperature method until we can think of a good check for it, it will just set the temperature to 440 K.

FoleyLab committed:
made some structural changes to pv_stpv methods

FoleyLab commented on wptherml/em.py:
```diff
@@ -955,65 +966,208 @@ def compute_stpv_gradient(self):
+            self._solar_spectrum * absorptivity_full_stack * env
+        )
+
+        self.pv_stpv_jsc = np.trapz(
```
Consider calling this attribute pv_stpv_short_circuit_current

FoleyLab commented on wptherml/em.py:
```diff
@@ -955,65 +966,208 @@ def compute_stpv_gradient(self):
+
+        # Iterate over material thicknesses
+        for i in range(0, e_gradient_index):
+            self.pv_stpv_short_circuit_current_gradient[i] = np.trapz(
```
Note this gradient is called pv_stpv_short_circuit_current_gradient - we should ensure that this name is matched to the pv-stpv short circuit current name as much as possible, so that it follows the pattern:
base_quantity_name
base_quantity_name_gradient

FoleyLab commented on wptherml/em.py:
```diff
@@ -955,65 +966,208 @@ def compute_stpv_gradient(self):
+            )  # Integrate for short circuit current
+
+    
+    def compute_pv_stpv_splitting_power(self):
```
This method needs to be completed.  I have added some notes to it; @AnonymousOceans please see if you can fill in the actual code before committing the changes to update the PR.  We can discuss if you need some assistance.

FoleyLab commented on wptherml/em.py:
```diff
@@ -955,65 +966,208 @@ def compute_stpv_gradient(self):
+    def compute_pv_stpv_splitting_power(self):
 
 
+    def compute_pv_stpv_total_incident_power(self):
```
This will be proportional to the short circuit current, and will also require information about the Fill Factor and Open-Circuit Voltage.  We do have that data, but let's leave dealing with this for later.

FoleyLab commented on wptherml/em.py:
```diff
@@ -955,65 +966,208 @@ def compute_stpv_gradient(self):
+
+        # get envelope function that behaves like ideal spectral response function
+        bg_idx = np.abs(self.wavelength_array - self.pv_lambda_bandgap).argmin()
+        env = np.zeros_like(self.wavelength_array)
```
Right now we are still using the "ideal" spectral response function.  That is fine for this stage, but let's not forget that we have started adding the required data to get a more realistic spectral response function.  After we merge this PR and try some optimizations and look at some data, we can go back and refine this part of the calculation.

FoleyLab reviewed (changes requested):
@AnonymousOceans I have made a few structural changes to this PR and also added some comments throughout to help guide your revisions to this PR.  Please look through them and let me know if you have any questions!

FoleyLab committed:
added comment to compute_pv_stpv_splitting_power

FoleyLab committed:
added pass statement to pv_stpv_splitting_power

AnonymousOceans committed:
Update em.py

Changed pv_stpv_jsc and pv_stpv_jsc_gradient to pv_stpv_short_circuit_current and pv_stpv_short_circuit_current_gradient.

AnonymousOceans committed:
Update em.py

Added 'else' statement in em.py for psc_thickness_option; set to 200, the optimal choice for the EQE arrays (from the given dataset).

AnonymousOceans committed:
Update em.py

Added a colon to an else statement

FoleyLab committed:
updated em.py to address revision request for PR #83

FoleyLab committed:
some old code made its way back in -fixed here

FoleyLab committed:
returned structures to original after jsc calcs

FoleyLab commented on wptherml/em.py:
```diff
@@ -1013,7 +1021,228 @@ def compute_pv_stpv(self):
+        
+    
+        # Store spectra 
+        emissivity_AR_polystyrene = self.emissivity_array
```
It is safer to copy numpy arrays using np.copy() rather than assign them using just an equals sign.

You can do this with the line
emissivity_AR_polystyrene = np.copy(self.emissivity_array)

FoleyLab replied:
You may (optionally) delete the lines where the transmissivity and reflectivity arrays are copied because we do not use those in this method

FoleyLab commented on wptherml/materials.py:
```diff
@@ -1643,3 +1643,71 @@ def _read_Atmospheric_Transmissivity(self):
+        if self.psc_thickness_option == 110:
+            _eqe_array = np.array([0, 0, 22, 46, 58, 59, 54, 40, 24, 18, 4,	0, 0])
+            
+        if self.psc_thickness_option == 200:
```
Remember to change the subsequent conditional statements to 'elif' statements instead of a series of 'if' statements.  Also conclude with an "else" statement that gives the matches the default value.

AnonymousOceans committed:
Update materials.py

Updated materials.py with elif statements and default else statement.

AnonymousOceans committed:
Update em.py

Added line
"emissivity_AR_polystyrene = np.copy(self.emissivity_array)"
as opposed to 
emissivity_AR_polystyrene = self.emissivity_array.
Removed unused reflectivity and transmissivity variables.

AnonymousOceans committed:
Update materials.py

Revised psc.thickness option if, elif, and else statement(s).

FoleyLab reviewed (approved):
Good to go!
